src/lenskit/sklearn/nmf.py

```diff
@@ -0,0 +1,99 @@
+# This file is part of LensKit.
+# Copyright (C) 2018-2023 Boise State University.
+# Copyright (C) 2023-2025 Drexel University.
+# Licensed under the MIT license, see LICENSE.md for details.
+# SPDX-License-Identifier: MIT
+
+"""
+This is for implicit
+
+This module contains a truncated SVD explicit-feedback scorer built on
+:class:`sklearn.decomposition.TruncatedSVD`.
+"""
+
+from __future__ import annotations
+
+import numpy as np
+from numpy.typing import NDArray
+from pydantic import BaseModel
+from sklearn.decomposition import non_negative_factorization
+from typing_extensions import Literal, override
+
+from lenskit.data import Dataset, ItemList, QueryInput, RecQuery
+from lenskit.data.vocab import Vocabulary
+from lenskit.logging import Stopwatch, get_logger
+from lenskit.pipeline import Component
+from lenskit.training import Trainable, TrainingOptions
+
+_log = get_logger(__name__)
+
+
+class NMFConfig(BaseModel, extra="forbid"):
+    beta_loss: Literal["frobenius", "kullback-leibler", "itakura-saito"] = "frobenius"
+    max_iter: int = 200
+    n_components: int | None = None
+    alpha_W: float = 0.0
+    alpha_H: float | Literal["same"] = "same"
+    l1_ratio: float = 0.0
+
+
+class NMFScorer(Component[ItemList], Trainable):
+    config: NMFConfig
+
+    users_: Vocabulary
+    items_: Vocabulary
+    user_components_: NDArray[np.float64]
+    item_components_: NDArray[np.float64]
```

Comment:
1. In newer code, I have been omitting the trailing `_`, and naming the component / embedding matrices `user_embeddings` and `item_embeddings`. So let's changes these (both the Vocabulary and NDArray attribute names) for consistency.
2. We have an `NPMatrix` type that is more precise than `NDArray`, let's use that instead as the type of the component declarations.
3. Let's use `np.float32`; you will need to use `np.require` in `train` to convert the component matrices to float32. This will save space in the final models (and 64-bit precision usually isn't needed anyway).